src/ZkFold/Base/Data/Vector.hs

```diff
@@ -5,110 +5,104 @@
-
-generate :: forall size a . KnownNat size => (Natural -> a) -> Vector size a
-generate f = Vector $ f <$> [0 .. value @size -! 1]
+unsafeToVector = Vector . V.fromList
 
 unfold :: forall size a b. KnownNat size => (b -> (a, b)) -> b -> Vector size a
-unfold f = Vector . ZP.take (value @size) . List.unfoldr (Just . f)
+unfold f = Vector . V.take (knownNat @size) . V.unfoldr (Just . f)
 
 fromVector :: Vector size a -> [a]
-fromVector (Vector as) = as
+fromVector (Vector as) = V.toList as
 
 (!!) :: Vector size a -> Natural -> a
-(Vector as) !! i = as List.!! fromIntegral i
+(Vector as) !! i = as V.! fromIntegral i
 
 uncons :: Vector size a -> (a, Vector (size - 1) a)
-uncons (Vector lst) = (P.head lst, Vector $ P.tail lst)
+uncons (Vector lst) = (V.head lst, Vector $ V.tail lst)
 
 reverse :: Vector size a -> Vector size a
-reverse (Vector lst) = Vector (P.reverse lst)
+reverse (Vector lst) = Vector (V.reverse lst)
 
 head :: Vector size a -> a
-head (Vector as) = P.head as
+head (Vector as) = V.head as
 
 tail :: Vector size a -> Vector (size - 1) a
-tail (Vector as) = Vector $ P.tail as
+tail (Vector as) = Vector $ V.tail as
 
 singleton :: a -> Vector 1 a
 singleton = Vector . pure
 
 item :: Vector 1 a -> a
-item (Vector [a]) = a
-item _            = error "Unreachable"
+item = head
 
 mapWithIx :: forall n a b . KnownNat n => (Natural -> a -> b) -> Vector n a -> Vector n b
-mapWithIx f (Vector l) = Vector $ zipWith f [0 .. (value @n -! 1)] l
+mapWithIx f (Vector l) = Vector $ V.zipWith f (V.enumFromTo 0 (value @n -! 1)) l
 
 mapMWithIx :: forall n m a b . (KnownNat n, Monad m) => (Natural -> a -> m b) -> Vector n a -> m (Vector n b)
-mapMWithIx f (Vector l) = Vector <$> M.zipWithM f [0 .. (value @n -! 1)] l
+mapMWithIx f (Vector l) = Vector <$> V.zipWithM f (V.enumFromTo 0 (value @n -! 1)) l
 
 zipWithM :: forall n m a b c . Applicative m => (a -> b -> m c) -> Vector n a -> Vector n b -> m (Vector n c)
-zipWithM f (Vector l) (Vector r) = Vector <$> M.zipWithM f l r
+zipWithM f (Vector l) (Vector r) = sequenceA . Vector $ V.zipWith f l r
 
 -- TODO: Check that n <= size?
 take :: forall n size a. KnownNat n => Vector size a -> Vector n a
-take (Vector lst) = Vector (ZP.take (value @n) lst)
+take (Vector lst) = Vector (V.take (knownNat @n) lst)
 
 drop :: forall n m a. KnownNat n => Vector (n + m) a -> Vector m a
-drop (Vector lst) = Vector (ZP.drop (value @n) lst)
+drop (Vector lst) = Vector (V.drop (knownNat @n) lst)
 
 splitAt :: forall n m a. KnownNat n => Vector (n + m) a -> (Vector n a, Vector m a)
-splitAt (Vector lst) = (Vector (ZP.take (value @n) lst), Vector (ZP.drop (value @n) lst))
+splitAt (Vector lst) = (Vector (V.take (knownNat @n) lst), Vector (V.drop (knownNat @n) lst))
 
 -- | The sole purpose of this function is to get rid of annoying constraints in ZkFols.Symbolic.Compiler.Arithmetizable
 --
 splitAt3 :: forall n m k a. (KnownNat n, KnownNat m) => Vector (n + m + k) a -> (Vector n a, Vector m a, Vector k a)
```

Comment:
If I remember correctly, this function is unused and we can remove it

Reply:
removed